### PR TITLE
Conform to the Datastore class

### PR DESCRIPTION
**Pull request by d2lam**
Modify the builds plugin to conform with the new Datastore

### CODE REVIEW

File: lib/save.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+'use strict';
+const Joi = require('joi');
+const DataModel = require('screwdriver-data-model');
+
+const IdModel = Joi.reach(DataModel.build.base, 'id');
+
+module.exports = (handler) => ({
+    method: 'POST',
```
Comment (FenrirUnbound) [inline lib/save.js]: If we're saving a specific build, shouldn't this be a PUT operation?

Comment (d2lam) [inline lib/save.js]: Probably should be [POST, PUT]. This would be used to create a build as well right?

File: test/index.test.js [hunk 2]
```diff
@@ -11,13 +11,15 @@ sinon.assert.expose(assert, { prefix: '' });
 
 describe('build plugin test', () => {
     const datastoreMock = {
-        scanAll: (cb) => { cb(); },
-        get: (id, cb) => { cb(null, id); }
+        save: (config, cb) => { cb(null); },
```
Comment (FenrirUnbound) [inline test/index.test.js]: This module has `sinon` included. Why are we choosing to not use `sinon` for a similar stubbing functionality?

File: test/index.test.js [hunk 3]
```diff
@@ -79,12 +81,55 @@ describe('build plugin test', () => {
...
+            datastoreMock.get.withArgs(config).yieldsAsync(null, testBuild);
             server.inject('/builds/1', (reply) => {
                 assert.equal(reply.statusCode, 200);
                 assert.deepEqual(reply.result, testBuild);
                 done();
             });
         });
     });
+
+    describe.only('/builds', () => {
```
Comment (FenrirUnbound) [inline test/index.test.js]: Remove the `.only`

File: test/index.test.js [hunk 4]
```diff
@@ -79,12 +81,55 @@ describe('build plugin test', () => {
...
+                        container: 'node:4'
+                    }
+                }
+            };
+
+            datastoreMock.save.withArgs(config).yieldsAsync(null);
+            const options = {
+                method: 'POST',
+                url: '/builds',
+                payload: JSON.stringify({
```
Comment (FenrirUnbound) [inline test/index.test.js]: `server.inject` will automagically stringify the object for you

File: lib/save.js (hunk 1 above)
Comment (FenrirUnbound) [inline lib/save.js]: Will the client be passing in the `id` every time?

Comment (d2lam) [inline lib/save.js]: Not sure, will probably discuss this tomorrow. How we want to auto-generate the id, etc.

Comment (FenrirUnbound) [inline lib/save.js]: If the `id` is not passed in, then you shouldn't need to reach for the `request.payload.id` value.

If the `id` is passed in, then it should be a PUT.

File: lib/list.js [hunk 5]
```diff
@@ -8,11 +8,20 @@ module.exports = (handler) => ({
...
+        description: 'Get builds with pagination',
         notes: 'Returns all build records',
         tags: ['api', 'builds'],
         handler: (request, reply) => {
-            handler.getBuilds(reply);
+            const config = {
+                table: 'builds',
+                params: {},
+                paginate: {
+                    page: request.query.page,
```
Comment (nkatzman) [inline lib/list.js]: are the query parameters required?

File: index.js [hunk 6]
```diff
@@ -12,34 +12,48 @@ let handler;
...
+ * @param  {Object}   config.params         Query parameters
+ * @param  {Object}   config.paginate       Pagination parameters
+ * @param  {Object}   config.paginate.count Number of items per page
+ * @param  {Object}   config.paginate.page  Specific page of the set to return
+ * @param  {Function} callback
  */
-Handler.prototype.getBuilds = function getBuilds(callback) {
-    this.datastore.scanAll(callback);
+Handler.prototype.getBuilds = function getBuilds(config, callback) {
+    this.datastore.scan(config, callback);
```
Comment (nkatzman) [inline index.js]: this needs the tableName

File: index.js [hunk 7]
```diff
@@ -12,34 +12,48 @@ let handler;
...
- * @param  {String}   id       Build identifier
- * @param  {Function} callback callback function
+ * @param  {Object}   config           Configuration
+ * @param  {String}   config.table     Table name
+ * @param  {String}   config.params    Query parameters
+ * @param  {String}   config.params.id Unique ID
+ * @param  {Function} callback
+ */
+Handler.prototype.getBuild = function getBuild(config, callback) {
+    this.datastore.get(config, callback);
```
Comment (nkatzman) [inline index.js]: this needs the tableName

File: index.js [hunk 8]
```diff
@@ -12,34 +12,48 @@ let handler;
...
+ * @param  {Object}   config           Configuration
+ * @param  {String}   config.table     Table name
+ * @param  {Object}   config.params    Record data
+ * @param  {Object}   config.params.id Unique ID
+ * @param  {Function} callback
  */
-Handler.prototype.getBuild = function getBuild(id, callback) {
-    this.datastore.get(id, callback);
+Handler.prototype.saveBuild = function saveBuild(config, callback) {
+    this.datastore.save(config, callback);
```
Comment (nkatzman) [inline index.js]: this needs the tableName

File: lib/list.js (hunk 5 above)
Comment (tkyi) [inline lib/list.js]: looks like they are expected
https://github.com/screwdriver-cd/screwdriver-datastore-imdb/blob/master/lib/imdb.js#L30

Comment (nkatzman) [inline lib/list.js]: What about requiring them for the http call?

File: lib/save.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+                    id: request.params.id,
+                    data: request.payload.data
+                }
+            };
+
+            handler.saveBuild(config, reply);
+        },
+        validate: {
+            payload: {
+                id: IdModel,
```
Comment (nkatzman) [inline lib/save.js]: should we remove this requirement?

File: lib/save.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+                data: Joi.object().required()
```
Comment (nkatzman) [inline lib/save.js]: Should this be the data model defined? https://github.com/screwdriver-cd/data-model/blob/master/model/build.js#L89

File: lib/list.js (hunk 5 above)
Comment (d2lam) [inline lib/list.js]: we should

File: lib/save.js (hunk 9 above)
Comment (d2lam) [inline lib/save.js]: yeAh 👍

File: lib/save.js (hunk 10 above)
Comment (d2lam) [inline lib/save.js]: Don't think so. base is everything. when we create we don't require all the fields.

File: lib/list.js (hunk 5 above)
Comment (d2lam) [inline lib/list.js]: @nkatzman  Added

**Comment by nkatzman** [comment]
👍

File: lib/list.js (hunk 5 above)
Comment (tkyi) [inline lib/list.js]: @stjohnjohnson added defaults

**Comment by FenrirUnbound** [comment]
👍 

ss0g2en

**Pull request merged**